core/src/cluster_slots_service/cluster_slots.rs

```diff
@@ -1,187 +1,337 @@
 use {
-    itertools::Itertools,
+    crate::replay_stage::DUPLICATE_THRESHOLD,
     solana_gossip::{
         cluster_info::ClusterInfo, contact_info::ContactInfo, crds::Cursor, epoch_slots::EpochSlots,
     },
-    solana_runtime::{bank::Bank, epoch_stakes::NodeIdToVoteAccounts},
-    solana_sdk::{
-        clock::{Slot, DEFAULT_SLOTS_PER_EPOCH},
-        pubkey::Pubkey,
-        timing::AtomicInterval,
-    },
+    solana_sdk::{clock::Slot, pubkey::Pubkey, timing::AtomicInterval},
     std::{
-        collections::{BTreeMap, HashMap},
-        sync::{Arc, Mutex, RwLock},
+        collections::{HashMap, VecDeque},
+        hash::RandomState,
+        ops::Range,
+        sync::{
+            atomic::{AtomicU64, Ordering},
+            Arc, Mutex, RwLock,
+        },
     },
 };
 
 // Limit the size of cluster-slots map in case
 // of receiving bogus epoch slots values.
-const CLUSTER_SLOTS_TRIM_SIZE: usize = 524_288; // 512K
+// This also constraints the size of the datastructure
+// if we are really really far behind.
+const CLUSTER_SLOTS_TRIM_SIZE: usize = 50000;
+
+pub type Stake = u64;
+
+//This is intended to be switched to solana_pubkey::PubkeyHasherBuilder
+type PubkeyHasherBuilder = RandomState;
+pub(crate) type ValidatorStakesMap = HashMap<Pubkey, Stake, PubkeyHasherBuilder>;
 
-pub(crate) type SlotPubkeys = HashMap</*node:*/ Pubkey, /*stake:*/ u64>;
+/// Pubkey-stake map for nodes that have confirmed this slot.
+/// If stake is zero the node has not confirmed the slot.
+pub(crate) type SlotPubkeys =
+    HashMap<Pubkey, AtomicU64 /* stake supporting */, PubkeyHasherBuilder>;
```

Comment:
The key datatype for this is a map from pubkeys to stakes, needs to be really fast and support atomic writes.